### ISSUE TITLE
token rejected with error invalid_grant Invalid JWT Signature

### ISSUE THREAD

**Issue by WellspringCS** [firebase/firebase-admin-dotnet#138]
Using Postman I get a token from Firebase via this endpoint:
https://identitytoolkit.googleapis.com/v1/accounts:signInWithPassword?key={{FirebaseApiKey}}

it returns to me an idToken: I've removed some characters from the below, but it's representative.

eyJhbGciOiJSUzI1NiIsImtpZCI6ImU1YT3RsbGMuY29tIl19LCJzaWduX2luX3Byb3ZpZGVyIjoicGFzc3dvcmQifX0.JljpJXibNtxrwnfb5OJzauOom-PgGSTFBw7eOl-zb6gj1jrYDzmbO9S4Y_dXegZY_LmjqnFC5Sf81PArDKvKH9P1CrJKoI_0yKJRk0lCFipsXWecE2jK3PG0qesjbPiEJv_BcO1Ia4R4r8JolJZnvt3Qg5q3f5WLOJpxxAA-2hD1Zp96cz5eV5d3fUiaHFFGvh0JC53vPnjjgzNooJu1949RYUHY2gwp0IleBKrjtRfOG92PkXqzWNLl3O5lWo9DBdrW8Wc6-IokuYLn4RfhbFYAysafSMkKjJ__rpNPmNFXYShO1nWYRNRkrPxnGmblfGGJcZTg2r7Am-ItdMd7Yg

Both my API and https://jwt.io/ are able to retrieve the user_id from the id_token, and then forward it on to FirebaseAuth.DefaultInstance.GetUserAsync(user_id);

I tried to validate the token via variations on this url
https://www.googleapis.com/oauth2/v1/tokeninfo?id_token=xxxxx
and not a one of them said the token was good.

Running on my local machine, this call succeeds and the API call is authorized. This uses a set of service account credentials locally stored.

Inside my GCP cluster, however, this call falls, with the below message:

Google.Apis.Auth.OAuth2.Responses.TokenResponseException: Error:"invalid_grant", Description:"Invalid JWT Signature.", Uri:"" at Google.Apis.Auth.OAuth2.Requests.TokenRequestExtenstions.ExecuteAsync(TokenRequest request, HttpClient httpClient, String tokenServerUrl, CancellationToken taskCancellationToken, IClock clock) at Google.Apis.Auth.OAuth2.ServiceAccountCredential.RequestAccessTokenAsync(CancellationToken taskCancellationToken) at Google.Apis.Auth.OAuth2.TokenRefreshManager.RefreshTokenAsync() at Google.Apis.Auth.OAuth2.TokenRefreshManager.ResultWithUnwrappedExceptions[T](Task`1 task) at Google.Apis.Auth.OAuth2.TokenRefreshManager.<>c.<GetAccessTokenForRequestAsync>b__10_0(Task`1 task) at System.Threading.Tasks.ContinuationResultTaskFromResultTask`2.InnerInvoke() at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)

I do not think this has to do with anything within my control, and this functionality was working recently.

**Comment by hiranya911** [comment]
Neither the ID toolkit REST API, nor the Google API client which seems to be throwing the above exception have anything to do with this repo. Please try asking on StackOverflow or reach out to GCP support. I'd also point you to https://github.com/googleapis/google-api-dotnet-client but without a more specific repro, I don't think they can help either.

All we can tell from the exception is that, some part of your application is trying to fetch an OAuth2 access token from a service account credential, but failing to do so.

**Comment by WellspringCS** [comment]
My apologies -- after posting it last night I decided the problem probably had to do not with the token, but with the credentials Firebase needs to log in and find info on the user. So I'll keep digging. Probably the error is mine.

Regardless, it appears I've picked the wrong Firebase audience, and I apologize for the distraction. 

But please help me understand. The error is happening in an **ASP.NET** API that has exactly one Firebase package:

```
<PackageReference Include="**FirebaseAdmin**" Version="1.9.1" />
```
It was on the basis of the above that I thought this was the forum for my question. I gather you're saying that firebase-admin-dotnet is ... not the above? That I'm really dealing with the google-api-dotnet-client?

**Comment by hiranya911** [comment]
@WellspringCS, The `FirebaseAdmin` package is indeed in this repo, but your exception has been thrown by the google-api-dotnet-client. The Google API client is responsible for fetching OAuth2 access tokens, and authorizing outgoing HTTP requests, but for some reason it's failing to do so in your app. On top of that you've also mentioned ID tokens and the Identity Toolkit REST API, which are whole other things. I don't see how they are related to either of the other 2 libraries involved here.

I think you should try to narrow down your issue a bit. See if you can come up with a minimal, independent repro. Since the error indicates a failure to fetch OAuth2 tokens from a service account, that's probably the first thing to check.

**Comment by WellspringCS** [comment]
OK, Thx - when I pick this up tomorrow, I'm going to look into the other credentials. I now suspect it's not the token that is the problem, but the credentials set up to initialize my FirebaseAdmin service. (And I'm assuming that the problem is mine.)

One question -- do you happen to know if there is a google service where one can throw a token at a google endpoint and verify if the token is valid? What threw me off last night (I'm now assuming that the token itself is fine) was google the endpoints I noted when I opened this issue. E.g.
https://www.googleapis.com/oauth2/v1/tokeninfo?id_token=xxxxx

They told me the token was invalid, and since I got the token from FirebaseAdmin endpoint, that's why I came here with my issue.

**Comment by hiranya911** [comment]
I'm not aware of a public endpoint that verifies ID tokens. The `/oauth2/v1/tokeninfo` seems to be related to OAuth2, and I wouldn't expect it to verify ID tokens. 

Now I haven't verified this. But at least in principle, if you're passing the `returnSecureToken: true` option to your `signInWithPassword` endpoint, it should send back an ID token that can be verified by the [`FirebaseAuth.VerifyIdTokenAsync()`](https://firebase.google.com/docs/reference/admin/dotnet/class/firebase-admin/auth/firebase-auth#verifyidtokenasync) API of the Admin SDK. But in the past I've also seen subtle differences among the tokens generated by the REST API, so this might not work. I can verify this tomorrow, if you decide to give that a try.

**Comment by WellspringCS** [comment]
> The /oauth2/v1/tokeninfo seems to be related to OAuth2, and I wouldn't expect it to verify ID tokens.
Ah, I've seen the names of tokens used interchangeably enough times that I didn't look too closely at that. Thanks for clarifying.

To the point! Today I revisited my service token and confirmed (idiot me) that it was out of date. So it was not the firebase token that was the problem, but the service token I was using when interacting with google-api-dotnet-client. I do apologize for taking up your time. I've filed this away in my pile of things to check when this stuff doesn't work.
